Use different exit codes for critical errors vs configuration errors

Jackenmen opened the pull request:
### Description of the changes

Fixes #5069 
Adds new possible exit codes: `2` (invalid cli usage) and `78` (configuration error)
These are now used appropriately in the bot which allows performing a different action depending on this value.
I updated the systemd service to work better now that we use different exit codes for critical errors (where restart is *usually* desired) and invalid CLI usage/configuration errors (where restart should always be not desired).
According to launchd's (Mac's service manager) manual, there is no option to discern between different non-zero exit codes, it treats them all the same. This means that Mac's service configuration will continue to work the same as it did before, sadly with no improvements.

### Have the changes in this PR been tested?

Yes

Jackenmen committed:
Update systemd guide

Jackenmen committed:
Add new exit codes to ExitCodes enum

Jackenmen committed:
Updates all sys.exit() usage to use ExitCodes enum and proper code

Jackenmen committed:
Move ExitCodes out of `redbot.core.bot`

Avoids circular import

Jackenmen committed:
Let's include Python version check in this change as well...

Jackenmen commented:
I went with restart *only* on 1 and 26 *but* I could alternatively do restart on *all* but 0, 2, and 78. The question is which one makes more sense here. It's not a given that we don't add more exit codes in the future and those codes would probably be non-restartable ones which would mean that the current state of this PR is preferred? I don't think adding *another* exit code on which we would want to restart is that likely in comparison.

Flame442 commented on redbot/core/cli.py:
```diff
@@ -25,7 +41,7 @@ def confirm(text: str, default: Optional[bool] = None) -> bool:
         except (KeyboardInterrupt, EOFError):
             print("\nAborted!")
-            sys.exit(1)
+            sys.exit(ExitCodes.INVALID_CLI_USAGE)
```
This doesn't match with other places where this error code is used. This is generally used for invalid usage of CLI flags, not the user ctrl c-ing. Other places which handle ctrl c, like when the user is prompted to enter their token, it uses `SHUTDOWN`. I think this makes more sense, since the user is explicitly trying to stop whatever is happening, even if it's done forcefully.

Jackenmen replied:
This is also a handler for `EOFError` which can occur in a non-interactive context where STDIN is closed. Therefore, it's possible that when used in a script, `redbot-setup` wasn't supposed to be aborted but due to incorrect usage it was. We could perhaps split this into two exception handlers instead though.

Flame442 commented on redbot/setup.py:
```diff
@@ -30,7 +31,7 @@
 except PermissionError:
     print("You don't have permission to write to '{}'\nExiting...".format(config_dir))
-    sys.exit(1)
+    sys.exit(ExitCodes.CRITICAL)
```
This should probably be `CONFIGURATION_ERROR`, since restarting won't give permission to write to this dir.

Jackenmen replied:
Yeah, probably makes more sense. I'm guessing I *probably* used `CRITICAL` because it concerns `redbot-setup` which isn't a daemon process (unlike `redbot`) anyway and you shouldn't restart it after failure.

It's also possible that I didn't want to use `CONFIGURATION_ERROR` in `redbot-setup` at all, maybe because places where it could make sense don't concern Red's configuration. I don't know...

I think `CONFIGURATION_ERROR` probably makes more sense here, I'll make the change.

Flame442 commented on redbot/setup.py:
```diff
@@ -77,7 +78,7 @@ def get_data_dir(*, instance_name: str, data_path: Optional[Path], interactive:
             " Provided path may contain an invalid character."
         )
-        sys.exit(1)
+        sys.exit(ExitCodes.CRITICAL)
```
This should probably be `CONFIGURATION_ERROR`, since restarting won't make the path valid.

Jackenmen replied:
I think this should be `INVALID_CLI_USAGE` as it concerns current user input rather than the existing configuration which is what all `CONFIGURATION_ERROR`s are for.

`CRITICAL` doesn't seem to make much sense here though, yeah.

Flame442 reviewed (changes requested):
A few of the exit codes used in certain places don't make sense to me, and I think they need a second look to see what they should be set to.

Jackenmen committed:
Address review

Jackenmen committed:
Stop mention running redbot-setup as admin...

Jackenmen commented:
Addressed the review, I've done one unrelated change to a message in redbot-setup failure mentioning running as admin because it bothered me:
[`85a4c47` (#5674)](https://github.com/Cog-Creators/Red-DiscordBot/pull/5674/commits/85a4c477fc6ba9720b0e1375a7bdc19459e752d6)

Flame442 reviewed (approved):
![](https://media.tenor.com/imR5A5P8_osAAAAd/christmas-kitty.gif)
